pkg/storegateway/bucket.go

```diff
@@ -1675,16 +1858,24 @@ func decodeSeries(b []byte, lsetPool *pool.SlabPool[symbolizedLabel], chks *[]ch
 
 		// Found a chunk.
 		if skipChunks {
-			// We are not interested in chunks and we know there is at least one, that's enough to return series.
-			return true, lset, nil
+			if streamingSeries {
+				// We are not interested in chunks, but we want the series to overlap with the query mint-maxt.
+				if maxt >= resMint && mint <= resMaxt {
+					// Chunk overlaps.
+					return true, lset, nil
+				}
+			} else {
+				// We are not interested in chunks and we know there is at least one, that's enough to return series.
+				return true, lset, nil
+			}
```

Comment:
i feel like this is a premature optimization and it affects the readability of `decodeSeries`, `unsafeLoadSeries`, and `loadingSeriesChunkRefsSetIterator`. There is already another type of filtering based on minT/maxT in `loadingSeriesChunkRefsSetIterator.symbolizedSet`.

We already use a single slice for all the chunk refs of all series (the `chks` here). This optimization only saves a copy of all chunk refs for all series. With the cache locality we have (`chks` is the same for all series; `ref`, `mint`, `maxt` are already in the L1 cache) i doubt this will be a problem

If you want to do this, can you run some benchmarks to prove it is necessary? I think you can modify `BenchmarkOpenBlockSeriesChunkRefsSetsIterator` to use streaming and run it with this change in decodeSeries and without it

Reply:
I ran `BenchmarkOpenBlockSeriesChunkRefsSetsIterator` with these two settings:
* old: streaming=true, skipchunks=true (the one used in this PR to get only the series).
* new: streaming=true, skipchunks=false (where we get the chunk refs in this iteration, but it will be discarded later when we are sending the series to the queries but things have been already allocated).

Looks like collecting the chunk refs is not really a good idea here. Though fetching from the chunk store is not a problem, we end up creating a lot of garbage in the memory. IMO refactoring to not create so much garbage will be more messy than simply skipping the chunks here.

<details>
  <summary>Benchmark results:</summary>

```
benchmark                                                                                                       old ns/op      new ns/op      delta
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="X"-10                                      12096          12125          +0.24%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="X",j="foo"-10                              23241          23326          +0.37%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="X",j!="foo"-10                             23247          23280          +0.14%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j=~"XXX|YYY"-10                               1831           1844           +0.71%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j=~"X.+"-10                                   1798           1806           +0.44%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~"X|Y|Z"-10                                 31801          32220          +1.32%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1"-10                                      117956583      149263274      +26.54%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",j="foo"-10                              120650088      132003318      +9.41%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j="foo",n="1"-10                              120687583      132569385      +9.85%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",j!="foo"-10                             112375525      123679866      +10.06%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~".*"-10                                    1940138542     2516478250     +29.71%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~".+"-10                                    3233826334     3789926000     +17.20%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~"^.+$",j=~"X.+"-10                         7689470        7662531        -0.35%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~""-10                                      1488853209     1487096375     -0.12%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i!=""-10                                      3230059583     3773573459     +16.83%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~".*",j="foo"-10                      126732222      136985281      +8.09%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="X",i=~"^.+$",j="foo"-10                    23521          23474          -0.20%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~".*",i!="2",j="foo"-10               135809234      147870250      +8.88%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i!=""-10                                1381886667     1408360375     +1.92%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i!="",j="foo"-10                        1367504250     1377869292     +0.76%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i!="",j=~"X.+"-10                       7057928        7001588        -0.80%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i!="",j=~"XXX|YYY"-10                   15224          15113          -0.73%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~"X|Y|Z",j="foo"-10                   55297          55102          -0.35%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~".+",j="foo"-10                      1365239333     1381735625     +1.21%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~"1.+",j="foo"-10                     160466125      161208036      +0.46%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~".+",i!="2",j="foo"-10               1376574459     1400964458     +1.77%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~".+",i!~"2.*",j="foo"-10             1507172584     1606615834     +6.60%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="X",i=~"^.+$",i!~"^.*2.*$",j="foo"-10       23635          23614          -0.09%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~"0xxx|1xxx|2xxx"-10                        254188         272885         +7.36%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~"(0|1|2)xxx"-10                            253907         271486         +6.92%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~"[0-2]xxx"-10                              254310         272172         +7.02%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i!~[0-2]xxx-10                                1957298750     2512282125     +28.35%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~".*",_i!~[0-2]xxx-10                       1917071750     2464767167     +28.57%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~"<unique_prefix>.+"-10                     99592          104492         +4.92%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~"<unique_prefix>.+"-10               1333923        1336715        +0.21%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i!~"<unique_prefix>.+"-10               118388648      142512349      +20.38%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j="foo",p="foo",i=~"<unique_prefix>.+"-10     12346222       12337420       -0.07%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j="foo",n=~".+",i=~"<unique_prefix>.+"-10     41176625       41592565       +1.01%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j="foo",n=~".*",i=~"<unique_prefix>.+"-10     11633155       11655490       +0.19%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j=~".*",n=~".*",i=~"<unique_prefix>.+"-10     126633         131757         +4.05%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j=~".+",n=~".+",i=~"<unique_prefix>.+"-10     57656463       57644575       -0.02%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/p!=""-10                                      417966833      521096396      +24.67%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="X"-10                                         12335          12262          -0.59%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="X",j="foo"-10                                 23483          23369          -0.49%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="X",j!="foo"-10                                23518          23452          -0.28%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/j=~"XXX|YYY"-10                                  1845           1837           -0.43%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/j=~"X.+"-10                                      1819           1815           -0.22%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/i=~"X|Y|Z"-10                                    31991          31766          -0.70%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1"-10                                         119572787      140986781      +17.91%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",j="foo"-10                                 120490921      133109271      +10.47%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/j="foo",n="1"-10                                 121290148      133251672      +9.86%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",j!="foo"-10                                112280014      123494139      +9.99%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/i=~".*"-10                                       1953297334     2511868125     +28.60%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/i=~".+"-10                                       3296579667     3827710083     +16.11%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/i=~"^.+$",j=~"X.+"-10                            7861789        7737615        -1.58%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/i=~""-10                                         1495865875     1486497083     -0.63%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/i!=""-10                                         3264578750     3796437334     +16.29%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",i=~".*",j="foo"-10                         127244375      136713896      +7.44%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="X",i=~"^.+$",j="foo"-10                       25746          23502          -8.72%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",i=~".*",i!="2",j="foo"-10                  136027505      148643607      +9.27%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",i!=""-10                                   1383761084     1400872958     +1.24%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",i!="",j="foo"-10                           1368707959     1380724458     +0.88%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",i!="",j=~"X.+"-10                          7040157        7048672        +0.12%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",i!="",j=~"XXX|YYY"-10                      15363          15179          -1.20%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",i=~"X|Y|Z",j="foo"-10                      55402          55203          -0.36%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",i=~".+",j="foo"-10                         1369373250     1378236833     +0.65%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",i=~"1.+",j="foo"-10                        160422762      160155304      -0.17%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",i=~".+",i!="2",j="foo"-10                  1381086042     1385087625     +0.29%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",i=~".+",i!~"2.*",j="foo"-10                1509172625     1519521667     +0.69%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="X",i=~"^.+$",i!~"^.*2.*$",j="foo"-10          23761          23537          -0.94%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/i=~"0xxx|1xxx|2xxx"-10                           255374         272551         +6.73%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/i=~"(0|1|2)xxx"-10                               254971         272581         +6.91%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/i=~"[0-2]xxx"-10                                 254271         272191         +7.05%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/i!~[0-2]xxx-10                                   1962049417     2469637542     +25.87%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/i=~".*",_i!~[0-2]xxx-10                          1919684667     2463027959     +28.30%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/i=~"<unique_prefix>.+"-10                        97770          104754         +7.14%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",i=~"<unique_prefix>.+"-10                  1333866        1334690        +0.06%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",i!~"<unique_prefix>.+"-10                  118268546      141630359      +19.75%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/j="foo",p="foo",i=~"<unique_prefix>.+"-10        12366665       12323407       -0.35%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/j="foo",n=~".+",i=~"<unique_prefix>.+"-10        41139698       41550463       +1.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/j="foo",n=~".*",i=~"<unique_prefix>.+"-10        11623795       11704082       +0.69%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/j=~".*",n=~".*",i=~"<unique_prefix>.+"-10        126419         132195         +4.57%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/j=~".+",n=~".+",i=~"<unique_prefix>.+"-10        57700270       57660614       -0.07%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/p!=""-10                                         415749542      520672542      +25.24%

benchmark                                                                                                       old allocs     new allocs     delta
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="X"-10                                      36             36             +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="X",j="foo"-10                              44             44             +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="X",j!="foo"-10                             44             44             +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j=~"XXX|YYY"-10                               31             31             +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j=~"X.+"-10                                   31             31             +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~"X|Y|Z"-10                                 46             46             +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1"-10                                      306675         906622         +195.63%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",j="foo"-10                              204003         503991         +147.05%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j="foo",n="1"-10                              204008         504024         +147.06%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",j!="foo"-10                             204020         504061         +147.06%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~".*"-10                                    5283238        20282782       +283.91%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~".+"-10                                    8032725        23033093       +186.74%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~"^.+$",j=~"X.+"-10                         100043         100043         +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~""-10                                      2750179        2750177        -0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i!=""-10                                      8033185        23032665       +186.72%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~".*",j="foo"-10                      204053         504015         +147.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="X",i=~"^.+$",j="foo"-10                    44             44             +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~".*",i!="2",j="foo"-10               204107         504109         +146.98%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i!=""-10                                3056844        3656804        +19.63%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i!="",j="foo"-10                        2954120        3254140        +10.16%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i!="",j=~"X.+"-10                       100071         100071         +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i!="",j=~"XXX|YYY"-10                   61             61             +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~"X|Y|Z",j="foo"-10                   82             82             +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~".+",j="foo"-10                      2954190        3254123        +10.15%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~"1.+",j="foo"-10                     328506         361835         +10.15%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~".+",i!="2",j="foo"-10               2954215        3254231        +10.16%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~".+",i!~"2.*",j="foo"-10             3237096        3503693        +8.24%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="X",i=~"^.+$",i!~"^.*2.*$",j="foo"-10       44             44             +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~"0xxx|1xxx|2xxx"-10                        450            901            +100.22%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~"(0|1|2)xxx"-10                            448            900            +100.89%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~"[0-2]xxx"-10                              449            900            +100.45%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i!~[0-2]xxx-10                                5282919        20282474       +283.93%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~".*",_i!~[0-2]xxx-10                       5282589        20282585       +283.95%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~"<unique_prefix>.+"-10                     225            376            +67.11%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~"<unique_prefix>.+"-10               256            264            +3.12%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i!~"<unique_prefix>.+"-10               306712         906716         +195.62%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j="foo",p="foo",i=~"<unique_prefix>.+"-10     278            311            +11.87%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j="foo",n=~".+",i=~"<unique_prefix>.+"-10     1343           1404           +4.54%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j="foo",n=~".*",i=~"<unique_prefix>.+"-10     267            328            +22.85%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j=~".*",n=~".*",i=~"<unique_prefix>.+"-10     259            410            +58.30%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j=~".+",n=~".+",i=~"<unique_prefix>.+"-10     1384           1533           +10.77%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/p!=""-10                                      1133073        4132844        +264.75%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="X"-10                                         36             36             +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="X",j="foo"-10                                 44             44             +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="X",j!="foo"-10                                44             44             +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/j=~"XXX|YYY"-10                                  31             31             +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/j=~"X.+"-10                                      31             31             +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/i=~"X|Y|Z"-10                                    46             46             +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1"-10                                         306720         906666         +195.60%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",j="foo"-10                                 204016         503988         +147.03%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/j="foo",n="1"-10                                 203991         504035         +147.09%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",j!="foo"-10                                204064         504014         +146.99%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/i=~".*"-10                                       5283099        20283203       +283.93%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/i=~".+"-10                                       8032722        23033179       +186.74%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/i=~"^.+$",j=~"X.+"-10                            100043         100043         +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/i=~""-10                                         2750171        2750173        +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/i!=""-10                                         8033053        23033504       +186.73%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",i=~".*",j="foo"-10                         204018         504038         +147.06%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="X",i=~"^.+$",j="foo"-10                       44             44             +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",i=~".*",i!="2",j="foo"-10                  204111         504093         +146.97%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",i!=""-10                                   3056733        3656793        +19.63%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",i!="",j="foo"-10                           2954011        3254176        +10.16%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",i!="",j=~"X.+"-10                          100071         100071         +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",i!="",j=~"XXX|YYY"-10                      61             61             +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",i=~"X|Y|Z",j="foo"-10                      82             82             +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",i=~".+",j="foo"-10                         2954124        3254269        +10.16%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",i=~"1.+",j="foo"-10                        328506         361843         +10.15%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",i=~".+",i!="2",j="foo"-10                  2954184        3254153        +10.15%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",i=~".+",i!~"2.*",j="foo"-10                3237138        3503863        +8.24%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="X",i=~"^.+$",i!~"^.*2.*$",j="foo"-10          44             44             +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/i=~"0xxx|1xxx|2xxx"-10                           449            901            +100.67%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/i=~"(0|1|2)xxx"-10                               448            900            +100.89%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/i=~"[0-2]xxx"-10                                 448            900            +100.89%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/i!~[0-2]xxx-10                                   5283046        20282647       +283.92%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/i=~".*",_i!~[0-2]xxx-10                          5283126        20282117       +283.90%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/i=~"<unique_prefix>.+"-10                        225            376            +67.11%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",i=~"<unique_prefix>.+"-10                  256            264            +3.12%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",i!~"<unique_prefix>.+"-10                  306724         906724         +195.62%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/j="foo",p="foo",i=~"<unique_prefix>.+"-10        282            313            +10.99%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/j="foo",n=~".+",i=~"<unique_prefix>.+"-10        1342           1404           +4.62%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/j="foo",n=~".*",i=~"<unique_prefix>.+"-10        268            328            +22.39%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/j=~".*",n=~".*",i=~"<unique_prefix>.+"-10        259            410            +58.30%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/j=~".+",n=~".+",i=~"<unique_prefix>.+"-10        1385           1534           +10.76%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/p!=""-10                                         1132877        4132954        +264.82%

benchmark                                                                                                       old bytes      new bytes      delta
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="X"-10                                      2268           2268           +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="X",j="foo"-10                              2989           2989           +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="X",j!="foo"-10                             2989           2989           +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j=~"XXX|YYY"-10                               1880           1880           +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j=~"X.+"-10                                   1824           1824           +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~"X|Y|Z"-10                                 2971           2971           +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1"-10                                      132742731      151509993      +14.14%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",j="foo"-10                              86983003       96538690       +10.99%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j="foo",n="1"-10                              86567467       96676298       +11.68%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",j!="foo"-10                             87185480       96505245       +10.69%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~".*"-10                                    3042543208     3522323408     +15.77%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~".+"-10                                    3209885416     3687882952     +14.89%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~"^.+$",j=~"X.+"-10                         11210196       11210182       -0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~""-10                                      240401640      240397432      -0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i!=""-10                                      3209026024     3688794416     +14.95%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~".*",j="foo"-10                      90733584       99987798       +10.20%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="X",i=~"^.+$",j="foo"-10                    3164           3164           +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~".*",i!="2",j="foo"-10               91800336       101499453      +10.57%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i!=""-10                                334226280      352852880      +5.57%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i!="",j="foo"-10                        288949376      298083976      +3.16%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i!="",j=~"X.+"-10                       11211765       11211714       -0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i!="",j=~"XXX|YYY"-10                   3902           3901           -0.03%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~"X|Y|Z",j="foo"-10                   5374           5376           +0.04%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~".+",j="foo"-10                      289329048      297294576      +2.75%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~"1.+",j="foo"-10                     44797803       45891092       +2.44%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~".+",i!="2",j="foo"-10               289566800      299279304      +3.35%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~".+",i!~"2.*",j="foo"-10             298355464      306169088      +2.62%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="X",i=~"^.+$",i!~"^.*2.*$",j="foo"-10       3373           3373           +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~"0xxx|1xxx|2xxx"-10                        146241         161756         +10.61%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~"(0|1|2)xxx"-10                            145769         160588         +10.17%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~"[0-2]xxx"-10                              146084         160861         +10.12%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i!~[0-2]xxx-10                                3041103712     3519885112     +15.74%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~".*",_i!~[0-2]xxx-10                       3044491504     3524315960     +15.76%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~"<unique_prefix>.+"-10                     47346          52317          +10.50%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~"<unique_prefix>.+"-10               1331159        1334529        +0.25%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i!~"<unique_prefix>.+"-10               132634539      151720441      +14.39%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j="foo",p="foo",i=~"<unique_prefix>.+"-10     19327053       19332466       +0.03%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j="foo",n=~".+",i=~"<unique_prefix>.+"-10     60283019       60285756       +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j="foo",n=~".*",i=~"<unique_prefix>.+"-10     12913902       12913734       -0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j=~".*",n=~".*",i=~"<unique_prefix>.+"-10     54261          59079          +8.88%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j=~".+",n=~".+",i=~"<unique_prefix>.+"-10     79512556       79509336       -0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/without_index_cache/p!=""-10                                      627062162      721975056      +15.14%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="X"-10                                         2267           2267           +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="X",j="foo"-10                                 2988           2988           +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="X",j!="foo"-10                                2988           2988           +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/j=~"XXX|YYY"-10                                  1880           1880           +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/j=~"X.+"-10                                      1824           1824           +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/i=~"X|Y|Z"-10                                    2972           2971           -0.03%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1"-10                                         133275467      151503708      +13.68%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",j="foo"-10                                 86435229       96304487       +11.42%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/j="foo",n="1"-10                                 86901335       96432249       +10.97%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",j!="foo"-10                                86962363       96478689       +10.94%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/i=~".*"-10                                       3044038016     3523771864     +15.76%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/i=~".+"-10                                       3209345600     3689194512     +14.95%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/i=~"^.+$",j=~"X.+"-10                            11210162       11210000       -0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/i=~""-10                                         240384872      240385032      +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/i!=""-10                                         3208955616     3688755784     +14.95%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",i=~".*",j="foo"-10                         90497175       99941880       +10.44%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="X",i=~"^.+$",j="foo"-10                       3165           3164           -0.03%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",i=~".*",i!="2",j="foo"-10                  91848495       101407026      +10.41%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",i!=""-10                                   334137640      352634560      +5.54%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",i!="",j="foo"-10                           288946256      297882472      +3.09%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",i!="",j=~"X.+"-10                          11211709       11211730       +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",i!="",j=~"XXX|YYY"-10                      3901           3901           +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",i=~"X|Y|Z",j="foo"-10                      5374           5374           +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",i=~".+",j="foo"-10                         288431688      298372936      +3.45%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",i=~"1.+",j="foo"-10                        44795275       45927109       +2.53%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",i=~".+",i!="2",j="foo"-10                  289470888      299875488      +3.59%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",i=~".+",i!~"2.*",j="foo"-10                298132696      306201032      +2.71%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="X",i=~"^.+$",i!~"^.*2.*$",j="foo"-10          3373           3373           +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/i=~"0xxx|1xxx|2xxx"-10                           146491         160562         +9.61%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/i=~"(0|1|2)xxx"-10                               145742         160570         +10.17%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/i=~"[0-2]xxx"-10                                 145388         160578         +10.45%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/i!~[0-2]xxx-10                                   3042054888     3521418848     +15.76%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/i=~".*",_i!~[0-2]xxx-10                          3043869536     3522842400     +15.74%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/i=~"<unique_prefix>.+"-10                        47301          52297          +10.56%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",i=~"<unique_prefix>.+"-10                  1332260        1335456        +0.24%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/n="1",i!~"<unique_prefix>.+"-10                  132577277      151595667      +14.35%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/j="foo",p="foo",i=~"<unique_prefix>.+"-10        19339242       19339386       +0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/j="foo",n=~".+",i=~"<unique_prefix>.+"-10        60277832       60283905       +0.01%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/j="foo",n=~".*",i=~"<unique_prefix>.+"-10        12914041       12912685       -0.01%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/j=~".*",n=~".*",i=~"<unique_prefix>.+"-10        54201          59187          +9.20%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/j=~".+",n=~".+",i=~"<unique_prefix>.+"-10        79512699       79512046       -0.00%
BenchmarkOpenBlockSeriesChunkRefsSetsIterator/with_index_cache/p!=""-10                                         626606450      722482484      +15.30%
``` 

</details>

Reply:
I agree allocating slices of chunk refs is not efficient, your benchmarks show it. What about doing the filtering in this function instead based on the returned `metas`?

https://github.com/grafana/mimir/blob/a51311ff38a96dbf79ee44d890260150e899bb7c/pkg/storegateway/series_refs.go#L956-L964


`metas` are already a single slice per batch, so it's not much garbage. But it avoid having to propagate three more args down to `decodeSeries`.